src/fdcache.cpp

```diff
@@ -739,6 +739,16 @@ bool FdManager::Close(FdEntity* ent, int fd)
                 // check another key name for entity value to be on the safe side
                 for(; iter != fent.end(); ){
                     if(iter->second.get() == ent){
+                        // [NOTE]
+                        // Since "ent" has already been freed just before this loop,
+                        // it must call "release()" here without being freed.
+                        // The return value of release() can be ignored, but clang-tidy
+                        // will warn.
+                        //
+                        const FdEntity* enttmp = iter->second.release();
```

Comment:
How do we have a situation where one `FdEntity*` is associated with different paths?  This seems like a scary situation since the owner of memory is the container.  Should this be a `shared_ptr` instead?

Reply:
I can't remember exactly, but I think it had something to do with renaming.
For example, I think it was something like renaming from another process while uploading a file.
I'll review this area after the next release. (I'm currently working on another improvement, and it will conflict a little with that, so I don't want to review it now.)

Reply:
https://github.com/s3fs-fuse/s3fs-fuse/pull/2532#issuecomment-2387518812 
This is one possible path

Reply:
if s3fs has two `unique_ptr` that have the same raw pointer then there is a problem.  The crash is telling us that an invariant is broken.  Manual `release` calls are a workaround for a logic bug.  I think we should remove the first path and `std::move` the `unique_ptr` before inserting the second path.  If it is actually required to have two paths with the same raw pointer then we should use a `shared_ptr` instead.

Reply:
I believe that when commit 1a50b9a04a82678b05e36927c323f42d94ca4a07 resolving the deadlock in ChangeEntityToTempPath, the scenario of using the tmp directory as cache was not taken into account. There shouldn't be two unique_ptrs pointing to the same raw pointer.

Reply:
@VVoidV @gaul Thank you for your comment.
My understanding was not sufficient, so I changed `unique_ptr` to `shared_ptr` and reviewed the code.
Please review again with the revised code.